### ISSUE TITLE
Zombie Processes

### ISSUE THREAD

**Issue by eljeffeg** [CybercentreCanada/assemblyline-v4-service#532]
When an updater fails to connect to a source, it doesn't handle the failure well.  It says it restarts the thread, but it doesn't properly kill the old thread leading to thousands of defunct processes and consuming resources.

<img width="1105" alt="Screenshot 2023-02-18 at 6 20 05 AM" src="https://user-images.githubusercontent.com/161904/219857971-efaf2ebf-0471-4819-abb0-f9fa9c115cf9.png">

From `al_svc_yara_updates`
```
2023-02-18 11:19:30,002  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | HTTPSConnectionPool(host='git.myrepo.net', port=443): Max retries exceeded with url: /yara/-/tree/master/signature-base (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable')))
2023-02-18 11:19:30,257  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | Restarting thread: Run source updates
2023-02-18 11:20:34,283  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | Repo clone failed with: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --branch=master -- https://*****:*****@git.myrepo.net/yara/-/tree/master/signature-base /tmp/tmp1tzmkudn/Yara_Rules
  stderr: 'Cloning into '/tmp/tmp1tzmkudn/Yara_Rules'...
fatal: unable to access 'https://service_assemblyline:Prohibition!@git.myrepo.net/yara/-/tree/master/signature-base/': Received HTTP code 503 from proxy after CONNECT
'
2023-02-18 11:21:34,077  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | HTTPSConnectionPool(host='git.myrepo.net', port=443): Max retries exceeded with url: /yara/-/tree/master/signature-base (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable')))
2023-02-18 11:21:34,308  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | Restarting thread: Run source updates
2023-02-18 11:22:39,216  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | Repo clone failed with: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --branch=master -- https://*****:*****@git.myrepo.net/yara/-/tree/master/signature-base /tmp/tmpzmykbrg6/Yara_Rules
  stderr: 'Cloning into '/tmp/tmpzmykbrg6/Yara_Rules'...
fatal: unable to access 'https://service_assemblyline:Prohibition!@git.myrepo.net/yara/-/tree/master/signature-base/': Received HTTP code 503 from proxy after CONNECT
'
2023-02-18 11:23:39,086  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | HTTPSConnectionPool(host='git.myrepo.net', port=443): Max retries exceeded with url: /yara/-/tree/master/signature-base (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 503 Service Unavailable')))
2023-02-18 11:23:40,338  WARNING 64e6fcacb1ea 1                assemblyline.updater.yara | Restarting thread: Run source updates
```

**Comment by eljeffeg** [comment]
One additional note, though this is specific to Yara as it has multiple downloads, is that a failure on one download will prevent the rest.  The expected behavior would be the other rules download and the service continues, just without the one rule that failed.

**Comment by cccs-sgaron** [comment]
@cccs-rs can you look at this please when you have time?

**Comment by cccs-rs** [comment]
Some recent PRs will improve the exception handling/logging for downloads. As for the defunct processes, I have a feeling the solution may be in core, in the ThreadedServerBase class by joining the threads once we know they've ended (although I'd want to test that to be sure.)

**Comment by cccs-rs** [comment]
This should be resolved on services re-built with: [v4.3.1.stable15](https://github.com/CybercentreCanada/assemblyline/releases/tag/v4.3.1.stable15)
